File: src/mpc_github_python/operations/actions.py
"""Actions module for GitHub workflow operations."""

from typing import Dict, List, Optional, Union
from dataclasses import dataclass
from pydantic import BaseModel
from github.Repository import Repository as GithubRepository
from github.Workflow import Workflow
from github.WorkflowRun import WorkflowRun
from github.GithubObject import NotSet

class WorkflowConfig(BaseModel):
    """Configuration for workflow operations."""
    name: str
    event: str
    ref: Optional[str] = None
    inputs: Optional[Dict[str, str]] = None

class WorkflowRunConfig(BaseModel):
    """Configuration for workflow run operations."""
    run_id: Union[str, int]

class ActionManager:
    """Manages GitHub workflow operations."""
    
    def __init__(self, repository: GithubRepository) -> None:
        """Initialize the action manager.
        
        Args:
            repository: GitHub repository to operate on
        """
        self.repository = repository
    
    async def get_workflow(self, workflow_id: int) -> Workflow:
        """Get a workflow by ID.
        
        Args:
            workflow_id: Workflow ID
            
        Returns:
            The workflow
        """
        return self.repository.get_workflow(workflow_id)
    
    async def list_workflows(self) -> List[Workflow]:
        """List all workflows in the repository.
        
        Returns:
            List of workflows
        """
        return list(self.repository.get_workflows())
    
    async def list_workflow_runs(
        self,
        workflow: Optional[Workflow] = None,
    ) -> List[WorkflowRun]:
        """List workflow runs.
        
        Args:
            workflow: Optional workflow to list runs for
            
        Returns:
            List of workflow runs
        """
        if workflow:
            return list(workflow.get_runs())
        return list(self.repository.get_workflow_runs())
    
    async def create_workflow_dispatch(
        self,
        workflow: Workflow,
        config: WorkflowConfig,
    ) -> Optional[WorkflowRun]:
        """Create a workflow dispatch event.
        
        Args:
            workflow: Workflow to dispatch
            config: Workflow configuration
            
        Returns:
            The workflow run if available
        """
        # Validate config
        config = WorkflowConfig.model_validate(config)
        
        ref = config.ref or self.repository.default_branch
        inputs = config.inputs or NotSet
        
        workflow.create_dispatch(
            ref=ref,
            inputs=inputs,
        )
        
        # Get the latest run
        runs = list(workflow.get_runs())
        return runs[0] if runs else None
    
    async def cancel_workflow_run(self, run: WorkflowRun) -> None:
        """Cancel a workflow run.
        
        Args:
            run: Workflow run to cancel
        """
        run.cancel()
    
    async def rerun_workflow_run(self, run: WorkflowRun) -> None:
        """Rerun a workflow run.
        
        Args:
            run: Workflow run to rerun
        """
        run.rerun()
    
    async def debug_workflow_run(
        self,
        config: WorkflowRunConfig,
    ) -> Dict[str, str]:
        """Debug a workflow run by getting its logs URL.
        
        Args:
            config: Workflow run configuration
            
        Returns:
            Dictionary containing the logs URL
        """
        # Validate config
        config = WorkflowRunConfig.model_validate(config)
        
        run = self.repository.get_workflow_run(config.run_id)
        if not run:
            raise ValueError(f"No workflow run found with ID {config.run_id}")
        
        logs_url = run.logs_url
        if not logs_url:
            raise ValueError("No logs found for this workflow run")
        
        return {"logs_url": logs_url} 

File: src/operations/actions.ts
import { z } from "zod";
import { githubRequest } from "../common/utils.js";

export const TriggerWorkflowOptionsSchema = z.object({
  owner: z.string(),
  repo: z.string(),
  workflow_id: z.union([z.string(), z.number()]),
  ref: z.string(),
  inputs: z.record(z.any()).optional()
});

/**
 * Triggers a GitHub Actions workflow dispatch event.
 */
export async function triggerWorkflow(params: z.TypeOf<typeof TriggerWorkflowOptionsSchema>): Promise<void> {
  const { owner, repo, workflow_id, ref, inputs } = TriggerWorkflowOptionsSchema.parse(params);
  await githubRequest(`https://api.github.com/repos/${owner}/${repo}/actions/workflows/${workflow_id}/dispatches`, {
    method: "POST",
    body: { ref, inputs }
  });
}

/**
 * Lists all workflows for a repository.
 */
export async function listWorkflows(owner: string, repo: string): Promise<any> {
  const response = await githubRequest(`https://api.github.com/repos/${owner}/${repo}/actions/workflows`);
  return response;
}

/**
 * Gets details of a specific workflow.
 */
export async function getWorkflow(owner: string, repo: string, workflow_id: string | number): Promise<any> {
  const response = await githubRequest(`https://api.github.com/repos/${owner}/${repo}/actions/workflows/${workflow_id}`);
  return response;
}

/**
 * Lists workflow runs. If workflow_id is provided, lists runs for that workflow; otherwise, lists all runs.
 */
export async function listWorkflowRuns(owner: string, repo: string, workflow_id?: string | number): Promise<any> {
  const endpoint = workflow_id ?
    `https://api.github.com/repos/${owner}/${repo}/actions/workflows/${workflow_id}/runs` :
    `https://api.github.com/repos/${owner}/${repo}/actions/runs`;
  const response = await githubRequest(endpoint);
  return response;
}

/**
 * Debug a failed workflow run by retrieving its logs URL.
 */
export const DebugWorkflowRunSchema = z.object({
  owner: z.string(),
  repo: z.string(),
  run_id: z.union([z.string(), z.number()])
});

export async function debugWorkflowRun(params: z.TypeOf<typeof DebugWorkflowRunSchema>): Promise<{ logs_url: string }> {
  const { owner, repo, run_id } = DebugWorkflowRunSchema.parse(params);
  const workflowRun = await githubRequest(`https://api.github.com/repos/${owner}/${repo}/actions/runs/${run_id}`);
  const runData = workflowRun as { logs_url?: string };
  if (!runData.logs_url) {
    throw new Error("No logs found for this workflow run.");
  }
  return { logs_url: runData.logs_url };
} 